### PR TITLE
Enable val/test loop disabling + datamodule tests

### PR DESCRIPTION
**Pull request by nateraw**
## What does this PR do?

This updates `MisconfigurationException`s to be warnings instead within trainer when you have dataloaders/defined steps for val/test that aren't configured correctly. Just because you have a validation dataloader doesn't mean you want to use it, and just because your model defines a val_step doesn't mean you should need to provide a validation dataset.

CC: @awaelchli we discussed something about this the other day 😄 

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by rohitgr7** [comment]
Nice! It solves the problem of commenting out the `validation_step` or `val_dataloader` / test code whenever it's not needed. 😅

**Comment by nateraw** [comment]
> Nice! It solves the problem of commenting out the `validation_step` or `val_dataloader` / test code whenever it's not needed. 😅

Exactly!! I have it a bit wrong so far here though. Test step should still raise misconfiguration but only if `self.testing`.

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1481,41 +1486,29 @@ def check_model_configuration(self, model: LightningModule):
...
-                    ' implement `test_dataloader` nor passed in `.test(test_dataloader)`.'
-                )
+            # Check test_dataloader, test_step and test_epoch_end
+            if self.is_overridden('test_dataloader', model):
+                if not self.is_overridden('test_step', model):
+                    rank_zero_warn(
+                        'You have passed in a `test_dataloader()`' ' but have not defined `test_step()`.',
+                    )
+            else:
+                if self.is_overridden('test_step', model):
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:1507]:
```suggestion
                if selt.testing and self.is_overridden('test_step', model):
```
I would suggest this to be called during testing only. Let's say `test_step` is implemented but I haven't implemented `test_dataloader` and doing `.fit()`.

**Comment by williamFalcon** [comment]
# Features:
- enables skipping val loop when not needed
- enables skipping test loop when not needed
- refactors some of the trainer
- adds datamodule tests
- adds config validator tests

File: pytorch_lightning/trainer/training_loop.py [hunk 2]
```diff
@@ -335,7 +335,9 @@ def train(self):
         # if reload_dataloaders_every_epoch, this is moved to the epoch loop
         if not self.reload_dataloaders_every_epoch:
             self.reset_train_dataloader(model)
-        self.reset_val_dataloader(model)
+
+        if model.val_dataloader is not None:
+            self.reset_val_dataloader(model)
```
Comment (awaelchli) [inline pytorch_lightning/trainer/training_loop.py:340]: it looks like val dataloaders get reloaded every epoch no matter the setting, `self.reload_dataloaders_every_epoch`. Isn't this a bug?

Comment (williamFalcon) [reply pytorch_lightning/trainer/training_loop.py:340]: that’s how it was before no? i think that was intended this way

Comment (williamFalcon) [reply pytorch_lightning/trainer/training_loop.py:340]: is there an issue about it?

Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:340]: if it's intended, ok. Just the name of the flag is a bit misleading then :)

> is there an issue about it?

There does not seem to be one.

**Pull request merged**